File: apps/api/src/services/social-security.service.ts
import { SocialSecurityFundType } from "@prisma/client";
import prisma from "../lib/prisma";
import { AppError } from "../middleware/errorHandler";
import { env } from "../config/env";

const FUND_TYPES: SocialSecurityFundType[] = [
  "EMERGENCY_HEALTH",
  "INSURANCE",
  "WELFARE",
  "EDUCATION",
  "RETIREMENT",
];

export async function getWorkerContributions(workerId: string): Promise<any[]> {
  const worker = await prisma.workerProfile.findUnique({
    where: { id: workerId },
  });

  if (!worker) {
    throw new AppError("Worker profile not found", 404);
  }

  let vaults = await prisma.socialSecurityVault.findMany({
    where: { workerId },
  });

  if (vaults.length === 0) {
    vaults = await Promise.all(
      FUND_TYPES.map((fundType) =>
        prisma.socialSecurityVault.create({
          data: {
            workerId,
            fundType,
            totalContributed: 0,
            employerMatch: 0,
            balance: 0,
            isOptedIn: false,
          },
        })
      )
    );
  }

  return vaults.map((v) => ({
    ...v,
    totalContributed: Number(v.totalContributed),
    employerMatch: Number(v.employerMatch),
    balance: Number(v.balance),
  }));
}

export async function toggleOptIn(
  workerId: string,
  fundType: SocialSecurityFundType,
  optedIn: boolean
): Promise<any> {
  const worker = await prisma.workerProfile.findUnique({
    where: { id: workerId },
  });

  if (!worker) {
    throw new AppError("Worker profile not found", 404);
  }

  if (!FUND_TYPES.includes(fundType)) {
    throw new AppError("Invalid fund type", 400);
  }

  let vault = await prisma.socialSecurityVault.findUnique({
    where: {
      workerId_fundType: { workerId, fundType },
    },
  });

  if (!vault) {
    vault = await prisma.socialSecurityVault.create({
      data: {
        workerId,
        fundType,
        totalContributed: 0,
        employerMatch: 0,
        balance: 0,
        isOptedIn: optedIn,
      },
    });
  } else {
    vault = await prisma.socialSecurityVault.update({
      where: { id: vault.id },
      data: { isOptedIn: optedIn },
    });
  }

  return {
    ...vault,
    totalContributed: Number(vault.totalContributed),
    employerMatch: Number(vault.employerMatch),
    balance: Number(vault.balance),
  };
}

export async function deductContribution(
  bookingId: string,
  workerId: string,
  workerPayoutAmount: number
): Promise<{ totalDeducted: number; deductions: any[] }> {
  const worker = await prisma.workerProfile.findUnique({
    where: { id: workerId },
  });

  if (!worker) {
    throw new AppError("Worker profile not found", 404);
  }

  const vaults = await prisma.socialSecurityVault.findMany({
    where: { workerId, isOptedIn: true },
  });

  if (vaults.length === 0) {
    return { totalDeducted: 0, deductions: [] };
  }

  const rate = env.SOCIAL_SECURITY_RATE;
  const deductions: any[] = [];
  let totalDeducted = 0;

  for (const vault of vaults) {
    const deductionAmount = Math.round(workerPayoutAmount * rate * 100) / 100;

    if (deductionAmount <= 0) continue;

    const employerMatch = Math.round(deductionAmount * 0.5 * 100) / 100;
    const newBalance =
      Number(vault.balance) + deductionAmount + employerMatch;
    const newTotalContributed =
      Number(vault.totalContributed) + deductionAmount;
    const newEmployerMatch = Number(vault.employerMatch) + employerMatch;

    await prisma.socialSecurityVault.update({
      where: { id: vault.id },
      data: {
        balance: newBalance,
        totalContributed: newTotalContributed,
        employerMatch: newEmployerMatch,
      },
    });

    await prisma.walletTransaction.create({
      data: {
        workerId,
        bookingId,
        type: "SOCIAL_SECURITY_DEDUCTION",
        amount: -deductionAmount,
        balanceAfter: Number(worker.walletBalance) - totalDeducted - deductionAmount,
        description: `Social security contribution: ${vault.fundType}`,
        reference: `SS-${vault.fundType}-${Date.now()}`,
      },
    });

    totalDeducted += deductionAmount;
    deductions.push({
      fundType: vault.fundType,
      deducted: deductionAmount,
      employerMatch,
      newBalance,
    });
  }

  return { totalDeducted, deductions };
}

export async function getContributionHistory(
  workerId: string,
  fundType: SocialSecurityFundType
): Promise<any> {
  const worker = await prisma.workerProfile.findUnique({
    where: { id: workerId },
  });

  if (!worker) {
    throw new AppError("Worker profile not found", 404);
  }

  const vault = await prisma.socialSecurityVault.findUnique({
    where: {
      workerId_fundType: { workerId, fundType },
    },
  });

  if (!vault) {
    throw new AppError("No vault found for this fund type", 404);
  }

  const transactions = await prisma.walletTransaction.findMany({
    where: {
      workerId,
      type: "SOCIAL_SECURITY_DEDUCTION",
      reference: { contains: fundType },
    },
    include: {
      booking: { select: { id: true, bookingRef: true } },
    },
    orderBy: { createdAt: "desc" },
  });

  return {
    vault: {
      ...vault,
      totalContributed: Number(vault.totalContributed),
      employerMatch: Number(vault.employerMatch),
      balance: Number(vault.balance),
    },
    transactions: transactions.map((t) => ({
      ...t,
      amount: Number(t.amount),
      balanceAfter: Number(t.balanceAfter),
    })),
  };
}

export default {
  getWorkerContributions,
  toggleOptIn,
  deductContribution,
  getContributionHistory,
};

File: backend/app/services/social_security_service.py
import time
import uuid

from datetime import datetime, timezone

from ..db import db, row_to_dict, rows_to_dicts
from ..errors import AppError
from ..utils import num, deep_serialize, now_utc

FUND_TYPES = ("EMERGENCY_HEALTH", "INSURANCE", "WELFARE", "EDUCATION", "RETIREMENT")


async def _ensure_vaults(worker_id: str) -> list:
    vaults = await db.fetch('SELECT * FROM "SocialSecurityVault" WHERE "workerId"=$1', worker_id)
    if not vaults:
        now = now_utc()
        for ft in FUND_TYPES:
            await db.execute(
                """
                INSERT INTO "SocialSecurityVault" (id, "workerId", "fundType", "totalContributed",
                    "employerMatch", balance, "isOptedIn", "createdAt", "updatedAt")
                VALUES ($1,$2,$3,0,0,0,false,$4,$4)
                """,
                str(uuid.uuid4()), worker_id, ft, now,
            )
        vaults = await db.fetch('SELECT * FROM "SocialSecurityVault" WHERE "workerId"=$1', worker_id)
    return vaults


async def get_worker_contributions(worker_id: str) -> list[dict]:
    worker = await db.fetchrow('SELECT * FROM "WorkerProfile" WHERE id=$1', worker_id)
    if worker is None:
        raise AppError("Worker profile not found", 404)
    vaults = await _ensure_vaults(worker_id)
    result = []
    for v in vaults:
        out = dict(v)
        out["totalContributed"] = num(v["totalContributed"])
        out["employerMatch"] = num(v["employerMatch"])
        out["balance"] = num(v["balance"])
        result.append(out)
    return deep_serialize(result)


async def toggle_opt_in(worker_id: str, fund_type: str, opted_in: bool) -> dict:
    worker = await db.fetchrow('SELECT * FROM "WorkerProfile" WHERE id=$1', worker_id)
    if worker is None:
        raise AppError("Worker profile not found", 404)
    if fund_type not in FUND_TYPES:
        raise AppError("Invalid fund type", 400)

    vault = await db.fetchrow(
        'SELECT * FROM "SocialSecurityVault" WHERE "workerId"=$1 AND "fundType"=$2',
        worker_id, fund_type,
    )
    now = now_utc()
    if vault is None:
        vault_id = str(uuid.uuid4())
        await db.execute(
            """
            INSERT INTO "SocialSecurityVault" (id, "workerId", "fundType", "totalContributed",
                "employerMatch", balance, "isOptedIn", "createdAt", "updatedAt")
            VALUES ($1,$2,$3,0,0,0,$4,$5,$5)
            """,
            vault_id, worker_id, fund_type, opted_in, now,
        )
    else:
        await db.execute(
            'UPDATE "SocialSecurityVault" SET "isOptedIn"=$1, "updatedAt"=$2 WHERE id=$3',
            opted_in, now, vault["id"],
        )
    vault = await db.fetchrow(
        'SELECT * FROM "SocialSecurityVault" WHERE "workerId"=$1 AND "fundType"=$2',
        worker_id, fund_type,
    )
    out = dict(vault)
    out["totalContributed"] = num(vault["totalContributed"])
    out["employerMatch"] = num(vault["employerMatch"])
    out["balance"] = num(vault["balance"])
    return deep_serialize(out)


async def get_contribution_history(worker_id: str, fund_type: str) -> dict:
    if fund_type not in FUND_TYPES:
        raise AppError("Invalid fund type", 400)
    worker = await db.fetchrow('SELECT * FROM "WorkerProfile" WHERE id=$1', worker_id)
    if worker is None:
        raise AppError("Worker profile not found", 404)
    vault = await db.fetchrow(
        'SELECT * FROM "SocialSecurityVault" WHERE "workerId"=$1 AND "fundType"=$2',
        worker_id, fund_type,
    )
    if vault is None:
        raise AppError("No vault found for this fund type", 404)
    rows = await db.fetch(
        """
        SELECT t.*, b."bookingRef" FROM "WalletTransaction" t
        LEFT JOIN "Booking" b ON b.id=t."bookingId"
        WHERE t."workerId"=$1 AND t.type='SOCIAL_SECURITY_DEDUCTION' AND t.reference LIKE $2
        ORDER BY t."createdAt" DESC
        """,
        worker_id, f"%{fund_type}%",
    )
    transactions = [{
        **dict(t), "amount": num(t["amount"]), "balanceAfter": num(t["balanceAfter"]),
        "booking": {"id": None, "bookingRef": t.get("bookingRef")},
    } for t in rows]
    vault_out = dict(vault)
    vault_out["totalContributed"] = num(vault["totalContributed"])
    vault_out["employerMatch"] = num(vault["employerMatch"])
    vault_out["balance"] = num(vault["balance"])
    return deep_serialize({"vault": vault_out, "transactions": deep_serialize(transactions)})
